src/array.cr

```diff
@@ -252,7 +252,7 @@ class Array(T)
   # ```
   #
   # See also: `#uniq`.
-  def &(other : Array(U)) forall U
+  def &(other : Array(U)) : Array(T) forall U
```

Comment:
Shouldn't the type actually be `Array(T & U)`?
We currently can't express that in the type grammar. But the implementation could use this type.

Reply:
No, it's always Array(T). It's the elements of self that are also in the given array.

Reply:
Ah, I guess you are right. But given that we can't express that... Or how would you change it?

Reply:
Changing this would bea breaking change

Reply:
No. Consider:
```
class Foo
  def ==(other)
    true
  end
end

x = [1,2,3]
y = [Foo.new]
p(x & y)
```